tests/gemm/test_mm_fp4.py

```
@@ -141,5 +141,58 @@ def test_mm_fp4_backend_auto(
+    )
+
+
+# K % 32 != 0 violates TMA 16-byte alignment; explicit b12x must reject cleanly.
+def test_mm_fp4_b12x_misaligned_k_raises():
+    cc = get_compute_capability(torch.device("cuda"))
+    if cc[0] != 12 or not (
+        torch.version.cuda and int(torch.version.cuda.split(".")[0]) >= 13
+    ):
+        pytest.skip("b12x backend requires SM120/SM121 + CUDA 13+.")
+    m, n, k = 64, 512, 112  # k % 32 == 16
+    input = torch.randn([m, k], device="cuda", dtype=torch.bfloat16)
```

Ruff (0.15.17)

[error] Variable `input` is shadowing a Python builtin

(A001)